tests/test_install_deps.py

```diff
@@ -23,42 +25,33 @@ def test_resolve_dependencies_install_deps__via_local_cache(local_pypi):
     }
 
 
-def test_resolve_dependencies_install_deps__handle_pip_install_failure(
+def test_resolve_dependencies_install_deps__raises_unresolved_error_on_pip_install_failure(
     caplog, local_pypi
 ):
-    # TemporaryPipInstallResolver will handle the inevitable pip install error
-    # and return to resolve_dependencies() with the missing package unresolved.
-    # For now, IdentityMapping "saves the day" and supplies a Package object.
-    # Soon, this should result in an unresolved package error instead.
+    # This tests the case where TemporaryPipInstallResolver encounters the
+    # inevitable pip install error and returns to resolve_dependencies()
+    # with the missing package unresolved.
+    # Since either install_deps or IdentityMapping are the final resolvers,
+    # this should raise an `UnresolvedDependenciesError`.
     caplog.set_level(logging.WARNING)
-    actual = resolve_dependencies(
-        ["does_not_exist"], setup_resolvers(install_deps=True)
-    )
-    assert actual == {
-        "does_not_exist": Package(
-            "does_not_exist", {"does_not_exist"}, IdentityMapping
-        ),
-    }
+
+    with pytest.raises(UnresolvedDependenciesError):
+        resolve_dependencies(["does_not_exist"], setup_resolvers(install_deps=True))
+
     assert all(word in caplog.text for word in ["pip", "install", "does_not_exist"])
 
 
-def test_resolve_dependencies_install_deps__pip_install_some_packages(
+def test_resolve_dependencies_install_deps_on_mixed_packages__raises_unresolved_error(
     caplog, local_pypi
 ):
-    debug_info = "Provided by temporary `pip install`"
-    caplog.set_level(logging.WARNING)
-    actual = resolve_dependencies(
-        ["click", "does_not_exist", "leftpadx"], setup_resolvers(install_deps=True)
+    caplog.set_level(logging.DEBUG)
+    deps = {"click", "does_not_exist", "leftpadx"}
+    # Attempting to pip install "does_not_exist"
+    # will result in an `UnresolvedDependenciesError`.
+    with pytest.raises(UnresolvedDependenciesError):
+        resolve_dependencies(deps, setup_resolvers(install_deps=True))
+    # Attempted to install deps with TemporaryPipInstall
+    assert (
+        f"Trying to resolve {deps!r} with <fawltydeps.packages.TemporaryPipInstallResolver"
+        in caplog.text
     )
-    # pip install is able to install "leftpadx", but "package_does_not_exist"
-    # falls through to IdentityMapping.
-    assert actual == {
-        "click": Package("click", {"click"}, TemporaryPipInstallResolver, debug_info),
-        "does_not_exist": Package(
-            "does_not_exist", {"does_not_exist"}, IdentityMapping
-        ),
-        "leftpadx": Package(
-            "leftpadx", {"leftpad"}, TemporaryPipInstallResolver, debug_info
-        ),
-    }
-    assert all(word in caplog.text for word in ["pip", "install", "does_not_exist"])
```

Comment:
I realize while reviewing this that the work we did to make `TemporaryPipInstallResolver` able to install as many packages as possible on a _best effort_ basis (when some packages fail to install) is currently not very useful: We always end up with `UnresolvedDependenciesError` and aborting in those cases. (That said the list of packages passed into `UnresolvedDependenciesError` is more accurate.)

I assure you that this is _not_ something I want to change now! I'm just starting to feel a small itch about maybe having the option to allow `TemporaryPipInstallResolver` and `IdentityMapping` at the same time. Let's see what our users say after we've released the mapping strategy milestone.

Reply:
Yes that is an unfortunate side-effect of raising an error on unresolved dependencies inside `resolve_dependencies()` itself. I wasn't able to assert that some dependencies were resolved with certain resolvers. I guess we can still assert that the error message refers to exactly the unresolved dependencies (which is in itself very useful data for the user to have)?

As for allowing identity mapping, I'm not convinced. If a dependency is not pip-installable, the best course of action to maintain "correctness" is either for the user to provide it in a local environment accessible to FawltyDeps or to provide it via a custom mapping.

Reply:
Agreed, I wonder if there is room for a test in between `test_resolve_dependencies_install_deps__raises_unresolved_error_on_pip_install_failure` and `test_resolve_dependencies_install_deps_on_mixed_packages__raises_unresolved_error`, to verify that when we install `{"click", "does_not_exist", "leftpadx"}`, the error message makes it clear that only `does_not_exist` failed, and not the two other packages.

I think your reasoning around the identity mapping makes sense as well. Let's wait and see what users think about the new mapping strategy before we make any more changes here.